sweepbatcher/presigned.go

```diff
@@ -0,0 +1,441 @@
+package sweepbatcher
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/btcsuite/btcd/blockchain"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+)
+
+// ensurePresigned checks that we can sign a 1:1 transaction sweeping the input.
+func (b *batch) ensurePresigned(ctx context.Context, newSweep *sweep) error {
+	if b.cfg.presignedHelper == nil {
+		return fmt.Errorf("presignedHelper is not installed")
+	}
+	if len(b.sweeps) != 0 {
+		return fmt.Errorf("ensurePresigned is done when adding to an " +
+			"empty batch")
+	}
+
+	sweeps := []sweep{
+		{
+			outpoint:  newSweep.outpoint,
+			value:     newSweep.value,
+			presigned: newSweep.presigned,
+		},
+	}
+
+	// Cache the destination address.
+	destAddr, err := b.getSweepsDestAddr(ctx, sweeps)
+	if err != nil {
+		return fmt.Errorf("failed to find destination address: %w", err)
+	}
+
+	// Set LockTime to 0. It is not critical.
+	const currentHeight = 0
+
+	// Check if we can sign with minimum fee rate.
+	const feeRate = chainfee.FeePerKwFloor
+
+	tx, _, _, _, err := constructUnsignedTx(
+		sweeps, destAddr, currentHeight, feeRate,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to construct unsigned tx "+
+			"for feeRate %v: %w", feeRate, err)
+	}
+
+	// Try to presign this transaction.
+	batchAmt := newSweep.value
+	signedTx, err := b.cfg.presignedHelper.SignTx(
+		ctx, tx, batchAmt, feeRate, feeRate,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to sign unsigned tx %v "+
+			"for feeRate %v: %w", tx.TxHash(), feeRate, err)
+	}
+
+	// Check the SignTx worked correctly.
+	err = CheckSignedTx(tx, signedTx, batchAmt, feeRate)
+	if err != nil {
+		return fmt.Errorf("signed tx doesn't correspond the "+
+			"unsigned tx: %w", err)
+	}
+
+	return nil
+}
+
+// presign tries to presign batch sweep transactions composed of this batch and
+// the sweep. It signs multiple versions of the transaction to make sure there
+// is a transaction to be published if minRelayFee grows.
+func (b *batch) presign(ctx context.Context, newSweep *sweep) error {
+	if b.cfg.presignedHelper == nil {
+		return fmt.Errorf("presignedHelper is not installed")
+	}
+	if len(b.sweeps) == 0 {
+		return fmt.Errorf("presigning is done when adding to a " +
+			"non-empty batch")
+	}
+
+	// Find the feerate needed to get into next block. Use conf_target=2,
+	nextBlockFeerate, err := b.wallet.EstimateFeeRate(ctx, 2)
```

Comment:
why is it 2? Maybe create a constant?

Reply:
I updated this place:
```go
        // priorityConfTarget defines the confirmation target for quick
        // inclusion in a block. A value of 2, rather than 1, is used to prevent
        // overestimation caused by temporary anomalies, such as periods without
        // blocks.
        const priorityConfTarget = 2

        // Find the feerate needed to get into next block.
        nextBlockFeeRate, err := b.wallet.EstimateFeeRate(
                ctx, priorityConfTarget,
        )
```

This fee rate is only used to define a high fee rate (10x of this value and at least 100 sats/vbyte).

Reply:
BTW the fee estimator fails with conf_target=1: https://github.com/lightninglabs/loop/issues/898
Maybe it worth making this a global constant in Loop and use it in all such places.

Reply:
Updated the comment:
```go
        // priorityConfTarget defines the confirmation target for quick
        // inclusion in a block. A value of 2, rather than 1, is used to prevent
        // fee estimator from failing.
        // See https://github.com/lightninglabs/loop/issues/898
        const priorityConfTarget = 2
```